### PR TITLE
Lifecycle staging 1

### PR DESCRIPTION
**Pull request by mjeronimo**
## Description of contribution in a few bullet points

This PR consists of the foundational modules for implementing lifecycle nodes in nav2, along with a couple simple modules that use the classes. The lifecycle work will be submitted in stages and this is the first PR in a series. 

The code won't be fully functional (and probably won't build) until all PRs in this series are reviewed, since this is a significant and incompatible change to most of the modules in the system. Perhaps we can review these changes and get approval, without merging until the whole set is OK'd.

### CODE REVIEW

**Comment by crdelsey** [comment]
@mjeronimo Are you planning to push all the PRs in the series or are you waiting on this to be reviewed before pushing the next?

**Comment by gbiggs** [comment]
What's going on with this `nav2_lifecycle` package? I'm wary of not pushing these changes upstream now rather than making them here. If navigation has its own lifecycle node classes there is a high risk that things will stay like that forever.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+class ILifecycle
+{
+public:
+  virtual ~ILifecycle() {}
+
+  // The basic lifecycle interface (minus onError and onShutdown, which are only relevant for nodes)
+  virtual CallbackReturn onConfigure(const rclcpp_lifecycle::State & state) = 0;
```
Comment (crdelsey) [inline nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp]: Method names are supposed to be either UpperCamelCase or snake_case according to the ROS2 style guide. I know we've been using lowerCamelCase a lot, but apparently that's not the correct style.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    const std::string & node_name,
+    const std::string & namespace_ = "",
+    bool use_rclcpp_node = false,
+    const rclcpp::NodeOptions & options = rclcpp::NodeOptions());
+  virtual ~LifecycleNode();
+
+protected:
+  // These two are recommended to be implemented by derived classes but have empty,
+  // default implementations
+  virtual CallbackReturn onShutdown(const rclcpp_lifecycle::State & state);
```
Comment (crdelsey) [inline nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp]: This is a bit confusing, having two different methods in the same class that just differ by naming style. Seems like it would be better if this included either `ILifecycle` or `LifecycleNode` using composition rather than inheritance.

File: nav2_lifecycle/package.xml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
```
Comment (crdelsey) [inline nav2_lifecycle/package.xml]: We're using format 3 everywhere now.

File: nav2_world_model/src/world_model.cpp [hunk 4]
```diff
@@ -12,67 +12,110 @@
...
+  auto size_y = costmap_->getSizeInCellsY();
+  auto data_length = size_x * size_y;
+  unsigned char * data = costmap_->getCharMap();
 
-  response->map.metadata.size_x = costmap_->getSizeInCellsX();
-  response->map.metadata.size_y = costmap_->getSizeInCellsY();
+  response->map.header.stamp = now();
+  response->map.header.frame_id = "map";
+  response->map.metadata.size_x = size_x;
+  response->map.metadata.size_y = size_y;
```
Comment (crdelsey) [inline nav2_world_model/src/world_model.cpp]: Do these changes have to do with Lifecycle nodes, or is this just some unrelated cleanup?

File: nav2_bt_navigator/package.xml [hunk 5]
```diff
@@ -8,7 +8,6 @@
   <license>Apache License 2.0</license>
 
   <buildtool_depend>ament_cmake</buildtool_depend>
-  <build_depend>nav2_common</build_depend>
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_bt_navigator/package.xml]: It was an integration error. Fixed.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 6]
```diff
@@ -28,13 +30,58 @@ namespace nav2_bt_navigator
...
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+BtNavigator::onCleanup(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "onCleanup");
+
+  task_server_->onCleanup(state);
+  task_server_.reset();
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: No, I don't think so. Typically, in the on_cleanup callback, the lifecycle node releases its resources (task_server_->reset(), in this case). The task_server itself also presents a lifecycle-style interface and must have its on_cleanup invoked. It's the mirror image of on_configure where the resource is allocated and then has on_configure invoked for it.

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 7]
```diff
@@ -55,21 +102,22 @@ BtNavigator::navigateToPose(const nav2_tasks::NavigateToPoseCommand::SharedPtr c
   blackboard->set<bool>("initial_pose_received", task_server_->isInitialPoseReceieved());  // NOLINT
 
   // Get the filename to use from the parameter
-  get_parameter_or<std::string>("bt_xml_filename", bt_xml_filename_,
+  std::string bt_xml_filename;
+  get_parameter_or<std::string>(std::string("bt_xml_filename"), bt_xml_filename,
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_bt_navigator/src/bt_navigator.cpp]: Yes, I agree. That needs to be changed, but I wasn't going to do it for this PR. It's important for non-determinism as you mention, but not strictly required to enable lifecycle management.

File: nav2_bt_navigator/src/main.cpp [hunk 8]
```diff
@@ -18,9 +18,13 @@
 
 int main(int argc, char ** argv)
 {
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_bt_navigator/src/main.cpp]: It was to disable buffering on stdout. I can remove these.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp]: Fixed.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+#define NAV2_LIFECYCLE__LIFECYCLE_HPP_
+
+#include "rclcpp_lifecycle/lifecycle_node.hpp"
+
+namespace nav2_lifecycle
+{
+
+using CallbackReturn = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn;
+
+class ILifecycle
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp]: Yes, agreed. I've proposed a LifecycleHelperInterface for rclcpp_lifecycle. For now, I'll rename this to LifecycleHelperInterface.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp (hunk 1 above)
Comment (mjeronimo) [reply nav2_lifecycle/include/nav2_lifecycle/lifecycle.hpp]: Yes, I'll update this code to match the lifecycle interface (on_configure, etc.).

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp (hunk 2 above)
Comment (mjeronimo) [reply nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp]: I'll fix the naming convention to match rclcpp_lifecycle::LifecycleNode and clean this up.

File: nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+  virtual CallbackReturn onError(const rclcpp_lifecycle::State & state);
+
+  // Overrides, for the standard rclcpp::LifecycleNode change_state service callbacks.
+  // These simply redirect to our own versions (OnConfigure, etc.) so that we can
+  // maintain a consistent naming convention. This also provides a degree of
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_lifecycle/include/nav2_lifecycle/lifecycle_node.hpp]: This comment will be removed as the differently-named methods will be removed.

File: nav2_lifecycle/package.xml (hunk 3 above)
Comment (mjeronimo) [reply nav2_lifecycle/package.xml]: Fixed in nav2_controller and nav2_lifecycle

File: nav2_lifecycle/src/lifecycle_node.cpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+: rclcpp_lifecycle::LifecycleNode(node_name, namespace_, options),
+  use_rclcpp_node_(use_rclcpp_node)
+{
+  if (use_rclcpp_node_) {
+    // Avoid the extra services we don't need for this node
+    rclcpp::NodeOptions local_node_options;
+    local_node_options.start_parameter_services(false);
+    local_node_options.start_parameter_event_publisher(false);
+
+    rclcpp_node_ = std::make_shared<rclcpp::Node>(node_name + "_rclcpp_node", namespace_,
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_lifecycle/src/lifecycle_node.cpp]: It's a convenience for the derived classes that require a normal ROS2 node in order to interface to classes that don't support LifecycleNodes yet (MessageFilter, TransformBroadcaster, TransformListener, etc.). The derived classes simply set a flag in the constructor and then have an rclcpp_node_ available for them to use in these cases. The alternative is to duplicate this code in all of the classes. I'd rather have it in one place. It's also simpler when the code will be removed. This file becomes obsolete and the parameter is simply removed from the contructor of the other classes (they will be creating rclcpp_lifecycle::LifecycleNode directly).

File: nav2_lifecycle/src/lifecycle_service_client.cpp [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  } while (rclcpp::ok() && status != rclcpp::executor::FutureReturnCode::SUCCESS);
+
+  return true;
+}
+
+const char *
+LifecycleServiceClient::transition_to_str(uint8_t transition)
+{
+  switch (transition) {
+    case Transition::TRANSITION_CREATE:
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_lifecycle/src/lifecycle_service_client.cpp]: It's actually not used anymore, so I'll remove it.
